### ISSUE TITLE
Minecarts and slopes.

### ISSUE THREAD

**Issue by C1ffisme** [minetest/minetest_game#1927]
![screenshot_20170113_125822](https://user-images.githubusercontent.com/9893266/32079259-9fe28fbc-ba6f-11e7-87e1-2572fba97b7e.png)

This doesn't just affect the First Person view:

![screenshot_20171026_170149](https://user-images.githubusercontent.com/9893266/32079300-c427b0fa-ba6f-11e7-9e47-94136e412c3c.png)

![screenshot_20171026_170150](https://user-images.githubusercontent.com/9893266/32079337-ea26eb18-ba6f-11e7-8d3e-c83b82da4946.png)

**Comment by paramat** [comment]
Yes i've noticed this too, I'm not too concerned about how other players look but the blocked view is annoying.
@SmallJoker

**Comment by C1ffisme** [comment]
I wonder if the player model can be rotated for the sake of making it look right to outside observers.

**Comment by paramat** [comment]
Pitch rotation? We don't have that ability.
Shifting the player to a new attachment position every time the cart is on a slope would solve both issues? Not sure if practical though.

**Comment by SmallJoker** [comment]
It's indeed possible to use the camera offset to work around the missing pitch rotation feature. I'll check what can be done.

**Comment by stujones11** [comment]
This is technically possible by attaching the player the actual 'Body' bone of the cart, however, some changes are required to the cart model to stop the player being flipped on its head on slopes. It is essentially the same hack I had to make for hand wielded items; because of what appears to be an Irrlicht bug, it is necessary to prevent the Body bone from ever being exactly at a rotation of 0,0,0, eg. 0.001,0,0 is perfectly fine and not at all noticeable. You then need to contend with the fact that the player is now permanently upside-down, you will either need to adjust the attachment rotation or flip the body bone of the model itself.

**Comment by DS-Minetest** [comment]
@stujones11 But attaching to a bone would also give the problem https://github.com/minetest/minetest/issues/2813.

**Comment by stujones11** [comment]
> But attaching to a bone would also give the problem

> because of what appears to be an Irrlicht bug, it is necessary to prevent the Body bone from ever being exactly at a rotation of 0,0,0

**Comment by InfiniteOtter** [comment]
The lazy method would be to have the mine cart's bin remain vertical while the carriage underneath rocks with the angle of the slope. 

Another way to make this work is to enlarge the cart so that the rider does not appear to pass through the walls of the cart.

**Comment by Ferk** [comment]
> You then need to contend with the fact that the player is now permanently upside-down, you will either have to invert the attachment rotation or the body bone of the model itself.

Would it be possible to add a new bone to the model for the player to attach to?
This has the benefit that you don't necessarily need the player to have the exact same angle as the cart, it would look more natural if the player is not radically turned 45° with every slope as if it was glued, just a little bit tilted and maybe with an offset to be a bit more in the center of cart, rather than having its feet merged with the cart wheels.

Also, in slopes you can see the rails through the cart, maybe the cart needs a bit of offset too.

**Comment by stujones11** [comment]
> Would it be possible to add a new bone to the model for the player to attach to?

That would possibly be the best compromise if we can be sure that bone attachment does not cause the player to lag behind the cart. I did not notice any when I tried it but I can confirm that attachment movement lagging is a real issue for objects attached to the player.

> If the height of the walls of the cart were shorter they wouldn't be so much in the way during slopes.

This seems like a good compromise, I honestly can't see constantly resetting the attachment rotation working well in multiplayer.

**Comment by Ferk** [comment]
> if we can be sure that bone attachment does not cause the player to lag behind the cart

Do you mean that there's a lag between the player and the cart animation?
I expect both animations would be synchronous, with the attachment being client-side and not really affected by network lag.

What I can imagine could happen is that the cart animation itself lags, simultaneously with the player. But I guess the lag in the cart animation happens regardless of the player being attached to it, though probably being attached makes it more noticeable.

**Comment by stujones11** [comment]
> Do you mean that there's a lag between the player and the cart animation in multiplayer?

The movement lagging I refer to only seems to effect specific bone attachment and does not occur when attaching to the object's 'root', ie `set_attach(parent, "", position, rotation)` It also appears to be much more prevalent for objects attached to the player rather than the other way, however, I am not sure how bad it can be. I am not really sure what causes this as you rightly pointed out, this should all be done client-side so may even be an Irrlicht issue.

Like I say, I did not see any movement lagging in the quick singleplayer test I made, though I do have a reasonably capable setup. I might investigate this matter further when I have time.

> EDIT: oh, I see you maybe meant the eye offset

Not really, that is another problem entirely, again you would expect it could be handled automatically client-side but I have no idea if that is possible with Irrlicht or not.

**Comment by paramat** [comment]
#1929 causes view weirdness that may need an engine fix.

**Comment by Wuzzy2** [comment]
FYI: The engine change is no longer needed. Entities can be rolled now.

**Comment by Wuzzy2** [comment]
Please rename this issue to something like "Bad camera/view when sitting in cart on sloped rail".
